doc/design/service-discovery.md

```diff
@@ -0,0 +1,421 @@
+
+InterfacePort caproSubscriber(interfacePortData);
+Listener myListener;
+
+void onDiscoveryUpdateCallback(InterfacePort* subscriber)
+{
+    subscriber->tryGetCaProMessage().and_then([](auto& caproMessage){
+        // Has any service that is relevant for me changed?
+        if(caproMessage.m_serviceDescription == myRelevantServiceDescription && someOtherCondition())
+        {
+            // call user-defined callback
+        }
+    });
+}
+
+myListener.attachEvent(caproSubscriber, DATA_RECEIVED, createNotificationCallback(onDiscoveryUpdateCallback));
+```
+
+#### Event-based notification: Alternative D
+
+```cpp
+class DiscoveryInfo
+{
+  public:
+
+    // Move all service-related methods from PoshRuntime to this class
+    cxx::expected<InstanceContainer, FindServiceError>
+    PoshRuntime::findService(const capro::ServiceDescription& serviceDescription) noexcept
+    {
+        // Update local service registry
+        get();
+
+        return filterFor(serviceDescription);
+    }
+
+    // @todo check if still needed
+    bool offerService(const capro::ServiceDescription& serviceDescription) noexcept;
+    void stopOfferService(const capro::ServiceDescription& serviceDescription) noexcept;
+
+    ServiceRegistryTopic get() noexcept
+    {
+        m_subscriber.take().and_then([](auto& ServiceRegistryTopic){
+            // Update our local copy of the service registry
+            m_lastServiceRegistry = ServiceRegistryTopic;
+        }).or_else([](){
+            // No change in service registry
+        });
+        return m_lastServiceRegistry;
+    }
+
+    ServiceRegistryTopic getDelta() noexcept
+    {
+        auto old = m_lastServiceRegistry;
+        get();
+        return old - m_lastServiceRegistry;
+    }
+
+    ServiceRegistryTopic getDelta(const capro::ServiceDescription& serviceDescription) noexcept;
+
+   filterFor(const capro::ServiceDescription& serviceDescription) noexcept;
+  private:
+    Subscriber<ServiceRegistryTopic> m_subscriber{"ServiceDiscovery", "GlobalInstance", "ServiceRegistryTopic"};
+    ServiceRegistryTopic m_lastServiceRegistry;
+}
+
+// Class should be attachable to Listener
+iox::popo::DiscoveryInfo discoveryInfo;
+myListener.attachEvent(discoveryInfo, DATA_RECEIVED, createNotificationCallback(userDefinedCallback));
+```
+
+#### Event-based notification: Alternative E
+
+```cpp
+// StatusPort implementation
+StatusPortHeader {
+    using uint64_t INVALID {0};
+    using uint64_t UPDATING {-1};
+    std::atomic<uint64_t> generation{INVALID};
+}
+template <typename T>
+using StatusPortData = Sample<T, StatusPortHeader>; // pretty neat, eh? We use a chunk to store the StatusPortData -> this means the number of ports and data sizes can be configured by the user without recompiling RouDi; could potentially also be done for the other ports
+
+template <typename T>
+StatusPortWriter {
+  public:
+    void write(cxx::function_ref<void(T&)> f);
```

Comment:
If the `StatusPortWriter` writes the data into the status port, why requires the lambda a non const `T` reference then? The writer only reads the data and writes it into the port?
If its the other way around the name would be somehow misleading or I am just confused.

Reply:
Code snippet is from @elBoberido my understanding: The writes the data via his lambda so it can't be const.

Reply:
The `T&` is the reference to th data wrapped by the `StatusPort` and the method would be used like this
```cpp
StatusPortWriter<uint64_t> status;
status.write([] (auto& s) {
    s = 42;
});
```